### PR TITLE
#1708 multiple filters + #2696 initial refactor

### PR DESCRIPTION
**Pull request by pkarw**
### Related issues:

#1708 
#2696 
#2682

### Short description and why it's useful

Multiple filters + URL deep linking - made by @BartoszLiburski, refactored by @pkarw; sub-filters are also refreshed when filtering out objects (layered navigation)
Planned for 1.10-rc1

@filrak and @patzick I won't be able (time constraints) to finish this PR if it will require any changes - please feel free to take care of it

It probably needs tons of testing - with the edge cases:
- if there are no-multiplied requests to `category/products` when switching the filters, or reloading the page
- to check if it still works with config `filters.multiSelect=false`
- TODO: the `beforeRouteEnter` need to be fixed - currently I've commented it all out as it interfered with filters switching

I've also refactored Category.js a bit - [pushing it towards the 2.0 architecture](https://github.com/DivanteLtd/vue-storefront/issues/2696)

### Screenshots of visual changes before/after (if there are any)

Screenshot 2019-04-14 00 33 07

### Which environment this relates to
Check your case. In case of any doubts please read about [Release Cycle](https://docs.vuestorefront.io/guide/basics/release-cycle.html)

- [x] Test version (https://test.storefrontcloud.io) - this is a new feature or improvement for Vue Storefront. I've created branch from `develop` branch and want to merge it back to `develop`
- [ ] RC version (https://next.storefrontcloud.io) - this is a stabilisation fix for Release Candidate of Vue Storefront. I've created branch from `release` branch and want to merge it back to `release`
- [ ] Stable version (https://demo.storefrontcloud.io) - this is an important fix for current stable version. I've created branch from `hotfix` or `master` branch and want to merge it back to `hotfix`

**IMPORTANT NOTICE** - Remember to update `CHANGELOG.md` with description of your change

### CODE REVIEW

File: core/compatibility/components/PriceSelector.js [hunk 1]
```diff
@@ -25,11 +25,16 @@ export default {
...
+    checked: {
+      type: Boolean,
+      required: false,
+      default: () => false
     }
   },
   data () {
     return {
-      active: false
+      active: this.checked
```
Comment (patzick) [inline core/compatibility/components/PriceSelector.js]: this should be removed. After change component should emit change and get state from `checked` prop

File: core/compatibility/components/PriceSelector.js [hunk 2]
```diff
@@ -44,15 +49,10 @@ export default {
     filterChanged (filterOption) {
       if (filterOption.attribute_code === this.code) {
         if (filterOption.id === this.id) {
-          if (this.active) {
-            this.active = false
-          } else {
-            this.active = true
-          }
+          this.active = !this.active
```
Comment (patzick) [inline core/compatibility/components/PriceSelector.js]: this should be an emit

File: core/compatibility/components/blocks/Category/Sidebar.js [hunk 3]
```diff
@@ -1,5 +1,6 @@
 import { buildFilterProductsQuery } from '@vue-storefront/core/helpers'
 import { mapGetters } from 'vuex'
+import rootStore from '@vue-storefront/core/store'
```
Comment (patzick) [inline core/compatibility/components/blocks/Category/Sidebar.js]: we should import config instead od store

File: core/modules/catalog/components/ProductCustomOption.ts [hunk 4]
```diff
@@ -20,11 +22,16 @@ export const ProductCustomOption = {
...
+    checked: {
+      type: Boolean,
+      required: false,
+      default: () => false
+    },    
   },
   data () {
     return {
-      active: false
+      active: this.checked
```
Comment (patzick) [inline core/modules/catalog/components/ProductCustomOption.ts]: we have here code redundncy, should be checked with price selector

File: core/modules/catalog/helpers/url.ts [hunk 5]
```diff
@@ -0,0 +1,27 @@
+export function buildFilterQueryString ({ route, filters }) {
```
Comment (patzick) [inline core/modules/catalog/helpers/url.ts]: i believe we should use `query-string` library for that - to check

File: core/modules/catalog/store/category/actions.ts [hunk 6]
```diff
@@ -200,10 +199,144 @@ const actions: ActionTree<CategoryState, RootState> = {
       }
     })
   },
+  pullMoreProducts(context, { route, store }) {
```
Comment (patzick) [inline core/modules/catalog/store/category/actions.ts]: check this and try to remove filters from state as they are in path now

File: core/modules/catalog/store/category/index.ts [hunk 7]
```diff
@@ -10,6 +10,11 @@ export const categoryModule: Module<CategoryState, RootState> = {
   state: {
     list: [],
     current: {},
+    pagination: {
```
Comment (patzick) [inline core/modules/catalog/store/category/index.ts]: should be also in route path with filters

Comment (pkarw) [reply core/modules/catalog/store/category/index.ts]: URL deep linking is an optional feature, moreover having filters in URLs is just a kind of “view” where the Vue filters is kind of “model” for rendering the query string; we shouldn’t remove it

File: core/pages/Category.js [hunk 8]
```diff
@@ -148,8 +114,21 @@ export default {
...
     }
   },
-  beforeRouteUpdate (to, from, next) {
-    this.validateRoute(to)
+  async beforeRouteUpdate (to, from, next) {
+    if (from.path !== to.path) { // category changed - not just filters
+      try {
+        await this.$store.dispatch('category/fetchAsync', { route: to, context: null, slug: to.params.slug })
+      } catch (err) {
+        if (err.message.indexOf('query returned empty result') > 0) {
```
Comment (Jensderond) [inline core/pages/Category.js]:
```suggestion
        if (err.message.indexOf('query returned empty result') >= 0) {
```
https://rules.sonarsource.com/java/tag/suspicious/RSPEC-2692

**Comment by pkarw** [comment]
I'm closing it in favor of #2890
